src/server/handlers/consents/{ID}/generateChallenge.ts

```diff
@@ -0,0 +1,130 @@
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Abhimanyu Kapur <abhi.kapur09@gmail.com>
+
+ --------------
+ ******/
+
+/* istanbul ignore file */
+
+/*
+ * This flag is to ignore BDD testing
+ * which will be addressed in the future in
+ * ticket #354
+ */
+
+import {
+  updateConsentCredential,
+  isConsentRequestInitiatedByValidSource,
+  putConsentId, ConsentCredential
+} from '../../../../domain/consents/{ID}/generateChallenge'
+import * as challenge from '../../../../lib/challenge'
+import { Request, ResponseToolkit, ResponseObject } from '@hapi/hapi'
+import { consentDB, scopeDB } from '../../../../lib/db'
+import { Consent } from '../../../../model/consent'
+import Logger from '@mojaloop/central-services-logger'
+import { convertScopesToExternal } from '../../../../lib/scopes'
+import { Scope } from '../../../../model/scope'
+
+// Asynchronously deals with generating challenge, updating consent db
+//  and making outgoing PUT consent/{ID} call
+export async function generateChallengeAndPutConsentId (
+  request: Request,
+  consent: Consent,
+  id: string
+): Promise<void> {
+  try {
+    // If there is no pre-existing challenge for the consent id
+    // Generate one and update database
+    if (!consent.credentialChallenge) {
+      // Challenge generation
+      const challengeValue = await challenge.generate()
+
+      // Updating credentials with generated challenge
+      const credential: ConsentCredential = {
+        credentialChallenge: challengeValue,
+        credentialStatus: 'PENDING',
+        credentialType: 'FIDO'
+      }
+
+      consent = await updateConsentCredential(consent, credential)
+    } else if (consent.credentialStatus === 'ACTIVE') {
+      // TODO: Error handling here - dealt with in #355
+      Logger.error('ACTIVE credential consent has requested challenge')
+      return
+    }
+
+    // Retrieve Scopes
+    const scopesRetrieved: Scope[] = await scopeDB.retrieveAll(id)
+    const scopes = convertScopesToExternal(scopesRetrieved)
+
+    // Outgoing call to PUT consents/{ID}
+    await putConsentId(consent, request, scopes)
+  } catch (error) {
+    Logger.push(error)
+    // eslint-disable-next-line max-len
+    Logger.error(`Error: Outgoing call with challenge credential NOT made to  PUT consent/${id}`)
+    // TODO: Decide on error handling HERE -  dealt with in future ticket #355
```

Comment:
Let rethrow the received error instead of keeping it silent. In the next story we will transfer it to PUT .../error

Reply:
@eoln if we throw an error here or in the previous comment - there will be an unhandled promise rejection issue as nothing is handling the POST handler's errors

Reply:
Yes, you are right, normally we shouldn't allow promise rejections to bubble up, but unhandled promise rejection will be temporary - until we got PUT .../error implemented. IMHO it is better to have comment TODO in code, and also unhandled promise rejection/exception which will help us remember that we have some unfinished flow

Reply:
makes sense have addressed this